runtime/ffi/ffi.cpp

```diff
@@ -0,0 +1,200 @@
+#include <cstring>
+#include <dlfcn.h>
+#include <ffi.h>
+#include <gmp.h>
+#include <stdexcept>
+
+#include "runtime/alloc.h"
+#include "runtime/header.h"
+
+extern "C" {
+
+#define KCHAR char
+#define TYPETAG(type) "Lbl'hash'" #type "{}"
+
+#define TAG_TYPE(NAME) static uint64_t tag_type_##NAME() {\
+  static uint64_t tag = -1; \
+  if (tag == -1) { \
+    tag = ((uint64_t)getTagForSymbolName(TYPETAG(NAME)) << 32) | 1; \
+  } \
+  return tag; \
+}
+
+  TAG_TYPE(void)
+  TAG_TYPE(uint8)
+  TAG_TYPE(sint8)
+  TAG_TYPE(uint16)
+  TAG_TYPE(sint16)
+  TAG_TYPE(uint32)
+  TAG_TYPE(sint32)
+  TAG_TYPE(uint64)
+  TAG_TYPE(sint64)
+  TAG_TYPE(float)
+  TAG_TYPE(double)
+  TAG_TYPE(uchar)
+  TAG_TYPE(schar)
+  TAG_TYPE(ushort)
+  TAG_TYPE(sshort)
+  TAG_TYPE(uint)
+  TAG_TYPE(sint)
+  TAG_TYPE(ulong)
+  TAG_TYPE(slong)
+  TAG_TYPE(longdouble)
+
+  mpz_ptr move_int(mpz_t);
+  char * getTerminatedString(string * str);
+
+  mpz_ptr hook_LIST_size(struct list * l);
+  block * hook_LIST_get(struct list * l, int idx);
+
+  static void * so_lib_handle() {
+    static void * handle = NULL;
+
+    if (handle == NULL) {
+      handle = dlopen(NULL, RTLD_LAZY);
+    }
+
+    return handle;
+  }
+
+  static ffi_type getTypeFromSymbol(uint64_t symbol) {
```

Comment:
this should return an ffi_type * (see comment below) obtained by getting the address of the constant struct.